### ISSUE TITLE
DagRun marked as success without executing task

### ISSUE THREAD

**Issue by aran3** [apache/airflow#17977]
### Apache Airflow version

2.1.3 (latest released)

### Operating System

Red Hat Enterprise Linux Server 7.6 (Maipo)

### Versions of Apache Airflow Providers

apache-airflow-providers-celery==2.0.0
apache-airflow-providers-ftp==2.0.0
apache-airflow-providers-http==2.0.0
apache-airflow-providers-imap==2.0.0
apache-airflow-providers-postgres==2.0.0
apache-airflow-providers-sqlite==2.0.0

### Deployment

Virtualenv installation

### What happened

Running a dag that is manually triggered, and has only 1 step, sometimes the scheduler marks the DAG as success without running the step at all. The step status remains Null in the db and in UI. 

### What you expected to happen

Scheduler should never mark Dag complete if it has "Null" state tasks. 
Probably this task should be marked as not "Null" before this function even runs, but I'm not sure about it. 

### How to reproduce

I don't have a fool-proof way to reproduce - but I was able to reproduce it when I brought down the scheduler for a long time and triggered the dag quickly a few times. 
It seems to happen more if dag is triggered multiple times quickly (in the order of seconds).

### Anything else

Happens about 20-50% of the time for this dag as described above. 

I reproduced it while running the scheduler with DEBUG, and the only relevant logs are:
```
[2021-09-01 21:58:17,358] {retries.py:78} DEBUG - Running SchedulerJob._create_dagruns_for_dags with retries. Try 1 of 3
[2021-09-01 21:58:17,365] {retries.py:78} DEBUG - Running SchedulerJob._get_next_dagruns_to_examine with retries. Try 1 of 3
[2021-09-01 21:58:17,376] {retries.py:78} DEBUG - Running SchedulerJob._get_next_dagruns_to_examine with retries. Try 1 of 3
[2021-09-01 21:58:17,402] {scheduler_job.py:1138} DEBUG - DAG feeds_dag not changed structure, skipping dagrun.verify_integrity
[2021-09-01 21:58:17,405] {dagrun.py:480} DEBUG - number of tis tasks for <DagRun feeds_dag @ 2021-09-01 20:15:07+00:00: mtf_9001_20210901_120__2021-09-01T20:15:07+00:00, externally triggered: True>: 0 task(s)
[2021-09-01 21:58:17,406] {dagrun.py:434} INFO - Marking run <DagRun feeds_dag @ 2021-09-01 20:15:07+00:00: mtf_9001_20210901_120__2021-09-01T20:15:07+00:00, externally triggered: True> successful
[2021-09-01 21:58:17,406] {dagrun.py:604} WARNING - Failed to record duration of <DagRun feeds_dag @ 2021-09-01 20:15:07+00:00: mtf_9001_20210901_120__2021-09-01T20:15:07+00:00, externally triggered: True>: start_date is not set.
```
Notice **0 task(s)**, which I expect is caused by the task being in Null state in DB and `DagRun.task_instance_scheduling_decisions` not checking for Null-state tasks. 

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

**Comment by enima2684** [comment]
Hi,
I am experiencing the same issue. 
I generally launch through the API many dagruns of the same dag (~365 dag runs). After running a few dagruns normally (~200), Airflow scheduler stops scheduling remaining runs and marks them as "succeded" even if no task was run

**Comment by hancyhancy123** [comment]
I am experiencing the same issue too. My task is executed every day. And if it can not run as usual some day, some other things will happen in my project. So I wonder if it can be discovered or inspected by some config, therefore I can trigger it manually and immediately once it happens next time.

**Comment by eladkal** [comment]
Does this still happens on latest Airflow version?

**Comment by rdgfuentes** [comment]
Hi. We are facing the same issue on v2.3.4+composer. Is there any plan on fixing this any time soon? Is there any configuration we could review/change to prevent or mitigate this? We have many DAGs that depends on others to run, when they are marked as success then the depending DAGs are being executed but they did not find any data because of this problem.

**Comment by enriquetuya** [comment]
Hi, we migrated from Google Cloud Platform Airflow 2.2.5+composer to Airflow 2.3.3+composer, and this change made our dags start failing with the description mentioned above. We have dynamic end dates,  based on several threads (https://stackoverflow.com/questions/73652663/airflow-dag-run-marked-as-success-although-the-tasks-didnt-run) seem to be causing this bug. We have more than 100 dags, we would like to know if there is a planned fix for this issue, to decide what we will do next. We've tried v2.3.4 and the issue persists.

**Comment by potiuk** [comment]
> Hi. We are facing the same issue on v2.3.4+composer. Is there any plan on fixing this any time soon? Is there any configuration we could review/change to prevent or mitigate this? We have many DAGs that depends on others to run, when they are marked as success then the depending DAGs are being executed but they did not find any data because of this problem.

> Hi, we migrated from Google Cloud Platform Airflow 2.2.5+composer to Airflow 2.3.3+composer, and this change made our dags start failing with the description mentioned above. We have dynamic end dates, based on several threads (https://stackoverflow.com/questions/73652663/airflow-dag-run-marked-as-success-although-the-tasks-didnt-run) seem to be causing this bug. We have more than 100 dags, we would like to know if there is a planned fix for this issue, to decide what we will do next. We've tried v2.3.4 and the issue persists.

Unless some of you submits additional information (logs, cicumstances, descriptions I don't think anything is going to happen - no-one will likely look at it, before we lack reproduction and information. The only way you can help us to speed up investigation and either fix, or helping you to diagnose bugs in your DAGs, is to provide more circumstancess - when it happens, examples of DAGs causing it (with specific information on how you generate start_date and the DAGs themselves). log files, observations, screenshots. 

This issue was not commented on since Sep 2021 and it was orignally reported in 2.1.3. Also, the stack overflow issue mentions that this might be due to dynamic ``start_date`` (which is bad practice and you should fix it in your DAGs - look at our current examples and https://airflow.apache.org/docs/apache-airflow/stable/best-practices.html). So most likely than not, the issue for people reporting it were not following the best practices. It is likely the same in your case and you should follow this trail I think.

Also, I am not sure where do you take from the information that there is a big and asking for "plans" to fix it. At this moment, we do not even know if this is a bug and certainly because of lack of that knowledge, there are no plans to fix it. 

Also be cautious about your expectations. You are paid Composer user, and while I understand you might have certain expectations and SLAs there, this is an OSS project that comes "as is" and therare not "plans" and "expectations" towards bug fixes of individual users. It does not work like that. Things get fixed when someone fixes them. And sometime things reported here are not real issues so they don't get fixed. Or sometimes it's unclear - and if there are no clarity those things won't get fixed either.

Some people here - users and maintainer - do their best to help users like you in their free time, but they do not provide any SLAs. Things are usually fixed if some member of the community (not necessarily maintainer) get enough information to diagnose them and we asses there are important to fix and decides to fix it. That's how the OSS works. 

So the best you can do (both of you @enriquetuya  and @rdgfuentes ) is to try to provide enough informaiton that people who try to help you for free, can actually help to solve your problem.

If you have some "expectations" and "SLA" - there are also companies and individuals who provide paid expertise for Airlfow - and maybe those will be better for you to approach if there are some expectations your business continuity depends on. That's much more "business" focused approach.

Help us to help you, please if you want help here.

**Comment by slumbi** [comment]
Happened to us as well on 2.3.4
What is strange that DagRun was in success state, but   the usual log ( "` Marking run <DagRun ...>  successful`" )  is **missing** from scheduler logs

**Comment by potiuk** [comment]
> Happened to us as well on 2.3.4 What is strange that DagRun was in success state, but the usual log ( "` Marking run <DagRun ...> successful`" ) is **missing** from scheduler logs

Still does not make it reproduceible case I am afraid.

**Comment by andres-arana** [comment]
@potiuk sorry for the delay on answering this. I'm part of the same team @enriquetuya and @rdgfuentes work on, so I'm talking on behalf of the team on this.

First, we understand the sorry reality of OSS development in which there's this huge number of entitled, mostly corporate users that demand updates, plans and SLA's from a project without giving anything back to the hundreds of people working on their free time to keep the project running. We are OSS maintainers in other projects as well and your message resonates with our experiences. We understand your message about being careful with our expectations, but I've re-read the messages from my coworkers multiple times and I think this is a misunderstanding. Maybe it's a language barrier problem as we are mostly spanish speakers, but when my coworkers were asking if there were any plans to fix this issue, it was not an expectation or a demand. We wanted to know if this is an known issue that someone was working on or if we still needed to provide more information. This was specifically asked because we knew this is a pretty old issue that might be completely unrelated to what we and some other users are experiencing on 2.3.4. In addition to this, this is the first time we are trying to debug / collaborate on a core airflow problem, so we are not sure how the project bug-fixing process works. Sorry if this came across wrong.

Now, the reason for the huge delay on the response is that we managed to fix this by switching to a single scheduler configuration. We are not sure why it fails consistently about 40% of the time on our setup when we enable multiple schedulers in composer and we have dags with dynamic end dates, and we are a bit lost on where to look. We think there's some sort of race condition (maybe transient state being saved to the db by one scheduler, and the other picking up right before the transition is done?).

1. Would you prefer we open a new issue to track this, since it most likely is a different issue than the one reported here? I think that would allow us to provide clearer information about our specific issue. Is GitHub issues the preferred place to create this or do you track issues somewhere else?
2. What kind of information would be useful to try and triage this? Again, this is the first time we try to collaborate on airflow core (we've worked on custom operators, plugins, etc. before, but never airflow core) and we are a bit lost on what would be useful.

Thank you, and again, sorry if this came across wrong.

**Comment by potiuk** [comment]
> First, we understand the sorry reality of OSS development in which there's this huge number of entitled, mostly corporate users that demand updates, plans and SLA's from a project without giving anything back to the hundreds of people working on their free time to keep the project running. 

No. It's not sorry. It's thriving, and fufliflls requirements and follows the lead of those who contribute. There is nothing sorry about it.


> We wanted to know if this is an known issue that someone was working on or if we still needed to provide more information.

Again - this is a complete misunderstanding of how it works here. this is not a corporate structure where you can ask one person "what's the state".  You have very wrong assumption about how things work in open-source project like this. There is not a single entity or person, that can answer it. If someone works on it, then that someone works on it. If somoene submits a fix, then someone submits a fix. And everyone else sees it. All discussions are public. And anyone - as any other person here can see it. There is not a single entity you can get an authoritative answer on whether someone is working on a random issue. No-one knows. No-one tracks it - other than what happens and is documented in issues like this one.


> This was specifically asked because we knew this is a pretty old issue that might be completely unrelated to what we and some other users are experiencing on 2.3.4. In addition to this, this is the first time we are trying to debug / collaborate on a core airflow problem, so we are not sure how the project bug-fixing process works. Sorry if this came across wrong.

It did not come "wrong". I have no "hard" feelings. I was **just** answering your questions. Simply answering the question "is someone working on it" by anyone (unless that person is working on fixing it) is impossible. Neither I, nor anyone else knows, nor track who is working on what and what's the state. This is a completely distributed way of work. Ther are some "projects" that get semi-organized, but on random issues like that one, there is no asnwer. It will get fixed when someone fixes it. this is the best answer anyone can get.


> 1. Would you prefer we open a new issue to track this, since it most likely is a different issue than the one reported here? I think that would allow us to provide clearer information about our specific issue. Is GitHub issues the preferred place to create this or do you track issues somewhere else?

If you think you can provide a reproducible case, then by all means open a new issue. Or provide mroe details on that one. Especially - if you migrate first to latest available version. Airlfow has hundreds bugs fixed every patchlevel release. The original issue has been reported on 2.1.3 whixh is > 1.5 yeear old and likely there are hundreds (if not thousands) bugs fixed since. And it's a complete loose of everyone's time to try to understand if the issue has been fixed since. It's way better - if you are really interested in fixing it - to upgrade to latest and see if you can still see it. So by all means - please, upgrade to latest airflow that you can and if you can reproduce it, open an issue describing it with all details you can muster.

> 2. What kind of information would be useful to try and triage this? Again, this is the first time we try to collaborate on airflow core (we've worked on custom operators, plugins, etc. before, but never airflow core) and we are a bit lost on what would be useful.
 
All.  That you can. Explain what investigations you've done so far, what you've tried and observations you had. Logs, circumstances, related logs from database around that time (if your investigations indicate they might be relevant). The thing is that when you report such errors, there is no "paid support" here. People who create airflow MIGHT help if they dedicate their own, personal time in trying to understand and diagnose YOUR problem. Which at the end might happen to be Airlfow problem, but just spending time on trying to understand it is usually unpaid time that people here might decide to spend time on. So do whatever you think is necessary to make this time as little as possible, so those people will eventually spend their time on solving your problem. But you have to understand that you are asking people for their fee time - to help your company to solve a commercial problem. So do whatever you think will make those people to want to spend their time.

One of the best ways to do it, is to show that you spend YOUR time, that you made effort to analyse the problem and tried to see what is wrong and that you tested and verified some hypotheses and you can share the result of that that will help to loose as little time on diagnosing as possible. this is the best way to make people want to help you - because they see this is not a demand but cooperation. 

There is no easy recipe for that. If we had an easy recipe, we would automate it. But usually problems like that have no easy recipes because you do not know what's wrong and it needs a human investing time and effort and looking at different hypotheses and providing useful insight.

Alternativelly - if you really WANT things to fix it badly - there are companies and individuals who provide paid support for Airlfow that you can pay for to get them look at your problem - usually that is connected with paying for their time and giving them access to your systems (possibly interactive with your employees that serve as "proxy' if the access cannot be given directly) to be able to analyse things.

> Thank you, and again, sorry if this came across wrong.

There is nothing to be sorry about. I do not feel anything was "wrong there". I just want to be plain and straightforward that any support here is not a "helpdesk" - it is provided on voluntary and "as-is" basis and it's you who ask for help, and you who should make all the effort that the ask for help might get successful.

**Comment by taehoka** [comment]
I still got same issue on airflow v2.4.3, Is there any updates regarding this?

**Comment by potiuk** [comment]
> I still got same issue on airflow v2.4.3, Is there any updates regarding this?

Well. It was also reported for 2.1.3.  The issue was closed because no activity from the user.  Airflow is Open Source Project - with no guarantees that all problems in all version will be addressed and fixed (in fact it comes with no guarantees at all if you look at the licence and the fact tha you pay exactly 0 for it). Many issues look similar to others but might be different. Many issues are solved without closing the issues because we have many 100s of isses, only volunteers working on it and we don't track which issues are solved by every single PR - that's why it has been closed due to inactivity.

But If you care about your issue being looked at, then the best thing you can do is to open new issue, where you decribe in detail what are YOUR problems, provide all teh evidences, and make yourself available as issue author to provide more evidences if your issue will be detailed enough that someone will decide to spend their time on looking at it (it's on you to provide enough evidences so that someone will have it easy to reproduce the issue and diagnose it).

So rather than asking about "update" on closed issue, I suggest you to take a more active approach and open issue, provide evidences and generally be proactive in making sure you follow up.

Additionally - I'd also suggest to upgrade Airflow - there are likely many 100s of issues solved even since 2.4.3 (we are at 2.7.3) so it's likely the problem you have is already fixed (and you miss on not having all those 100s of issues fixing also possibly other problems).
